stellargraph/layer/sort_pooling.py

```diff
@@ -0,0 +1,97 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2018-2020 Data61, CSIRO
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+from tensorflow.keras.layers import Layer
+from ..core.experimental import experimental
+
+
+@experimental(reason="Missing unit tests and generally untested.", issues=[1044])
+class SortPooling(Layer):
+
+    """
+    Sort Pooling Keras layer.
+
+    Original paper: An End-to-End Deep Learning Atchitecture for Graph Classification, M. Zhang, Z. Cui, M. Neumann, and
+    Y. Chen, AAAI-18, https://www.aaai.org/ocs/index.php/AAAI/AAAI18/paper/viewPaper/17146
+
+
+    Args:
+        k (int): The number of rows of output tensor.
+
+    """
+
+    def __init__(self, k):
+        super().__init__()
+
+        self.trainable = False
+        self.k = k
+
+    def get_config(self):
+        """
+        Gets class configuration for Keras serialization. Used by keras model serialization.
+
+        Returns:
+            A dictionary that contains the config of the layer
+        """
+        return {"k": self.k}
+
+    def compute_output_shape(self, input_shapes):
+        """
+        Computes the output shape of the layer.
+        Assumes the following inputs:
+
+        Args:
+            input_shapes (tuple of ints)
+                Shape tuples can include None for free dimensions, instead of an integer.
+
+        Returns:
+            An input shape tuple.
+        """
+        return input_shapes[0], self.k, input_shapes[2]
+
+    def call(self, inputs, **kwargs):
+        """
+        Applies the layer.
+
+        Args:
+            inputs (list): a list of 3 input tensors that includes
+                node features (size B x N x Sum F_i),
+
+                where B is the batch size, N is the number of nodes in the largest graph in the batch, and
+                F_i is the dimensionality of node features output from the i-th convolutional layer.
+
+        Returns:
+            Keras Tensor that represents the output of the layer.
+        """
+        outputs = tf.map_fn(
+            lambda x: tf.gather(
+                x, tf.argsort(x[..., -1], axis=0, direction="DESCENDING")
```

Comment:
The paper talks about using index `-2` (and `-3`, `-4`, ...) to break ties. This doesn't seem to be doing so, is that important?

(from https://github.com/stellargraph/stellargraph/pull/1212/files#r403892314)

Reply:
I'll just quote myself from above which also quotes from the paper,

> Note that sorting is performed using only the last column of the input tensor as stated in [1], "For convenience, we set the last graph convolution to have one channel and only used this single channel for sorting."

I will update the docstring to mention the above. Whether it makes a difference or not, I do not know since the original paper does not show results for sorting with other than the last column. I suspect that given enough GCN layers, the approximation is good enough and that there are few if any ties to break that, in practice, makes little difference to sort beyond the one column.